### PR TITLE
Fix a few decoding bugs, and increase test coverage.

### PR DESCRIPTION
**Pull request by plietar**
- The content-type parameter was decoded as QCBOR_TYPE_ANY, but then
  assumed to always be a string. This could cause a type confusion and
  lead to crashes. It is now changed to be accept strings on integers
  only.

- The handling of a missing alg header parameter was creating an
  exception but not throwing it. This didn't weaken validation, since a
  bad_optional exception would be thrown two lines further down instead,
  but produced confusing and unexpected error responses.

- Added tests for various encoding and invalid messages edge cases. We
  cannot rely on pycose for many of these as it refuses malformed
  messages, so we use cbor2 directly.

- When embedding a receipt in a Sign1 enveloppe, which contained a
  single unprotected X509 certificate x5chain, the wrong QCBOR function
  was used, producing incorrectly formatted outputs.

- Check the return value when decoding X509 certificates and throw a
  ValidationError. Previously this could cause NULL pointer
  dereferencing.

### CODE REVIEW

File: test/test_ccf.py [hunk 1]
```diff
@@ -172,6 +190,32 @@ def test_submit_claim_notary_x509(
...
+        client.submit_claim(claims)
+
+
+def test_root_identity(
+    client: Client,
+    trust_store: TrustStore,
+    trusted_ca: X5ChainCertificateAuthority,
+):
+    """
+    Submit a claim signed by the self-signed root certificate.
```
Comment (letmaik) [inline test/test_ccf.py:211]: A certificate with the CA bit set should never be used for anything other than signing certificates that are issued. The fact that we allow this is a bug. In the trust store, there should only be CA certificates or self-signed non-CA certificates.

**Comment by plietar** [comment]
Changes to X509 validation was broken off into PR #104.

**Comment by letmaik** [comment]
@plietar X.509 changes are still in here, right?

**Comment by plietar** [comment]
@letmaik it is based on the other PR, just because I've moved stuff a bit in the test infra, but it made Github display all of it. I've change the target branch so only the encoding fixes should be visible now.

**Review by letmaik — APPROVED** [review]
Should be merged into main once the other PR is in.

**Comment by plietar** [comment]
@letmaik made some minor new changes following the rebase on the x509 PR. Main one being that I realized the service would crash on a ill-formed DER in the x5chain. Fixed now.

**Review by letmaik — APPROVED** [review]
